Various solidcp updates

spardoko opened the pull request:
# Description

Various updates that mattered to me.

Fixes # (issue)

These commits have been tested and are used in my production environment.

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Please also note any relevant details for your test configuration.

- [ ] Test A
- [ ] Test B

- [ ] Built code to ensure it has no errors

# Checklist:

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] Any dependent changes have been merged and published in downstream modules

IvanSchulz commented on SolidCP/Sources/SolidCP.WebPortal/DesktopModules/SolidCP/VPS2012/VpsDetailsEditConfiguration.ascx:
```diff
@@ -226,7 +229,7 @@
 
                     <div class="text-right">
                         <CPCC:StyleButton id="btnCancel" CssClass="btn btn-warning" runat="server" CausesValidation="False" OnClick="btnCancel_Click"> <i class="fa fa-times">&nbsp;</i>&nbsp;<asp:Localize runat="server" meta:resourcekey="btnCancel"/> </CPCC:StyleButton>&nbsp;
-                        <CPCC:StyleButton id="btnUpdate" CssClass="btn btn-success" runat="server" meta:resourcekey="btnUpdate" OnClick="btnUpdate_Click" ValidationGroup="Vps" OnClientClick="if(!confirm('Before applying new configuration VPS could be stopped.\n\nAfter the configuration is changed it will be started again automatically.\n\nDo you want to proceed?')) return false; ShowProgressDialo ...
```
meta:resourcekey="btnUpdate" is used for localizing JavaScript dialog (btnUpdate.OnClientClick). There is no german confirm dialog when it is removed

IvanSchulz commented:
Hi @spardoko , thank you for the pull request.
The commit [added configurable limits to vps](https://github.com/FuseCP/SolidCP/pull/218/commits/a9af80c7fd8eb949f1ef985a413d6d27107ef0d8) shows the RAM and HDD limits as text only, but allows for overhead.
@FuseCP-TRobinson is this ok to use this settings in web.config? - https://github.com/FuseCP/SolidCP/pull/218/commits/10ebdd57ad6870d196aef1aebd653d3e6c313a68#diff-ffa8b3e557eb4a14702df897e78eed7a3665a3c1d98b58f58ed8dd2b92708b34
The another changes should be ok i think.

FuseCP-TRobinson commented:
> @FuseCP-TRobinson is this ok to use this settings in web.config?

Ideally we should be setting these via the provider settings not the web.config so it can be adjusted per server.

Settings options in web.config should only be done when its not possible via other methods(That does go for other settings). As we move toward the CoreWCF version we will be cleaning up those settings.

berkut1 commented:
I don't understand the point of this: 
[added configurable limits to vps](https://github.com/FuseCP/SolidCP/pull/218/commits/a9af80c7fd8eb949f1ef985a413d6d27107ef0d8)

Can’t we achieve this by extending quotas with add-ons in the existing system? I remember starting work on this but didn’t finish for reasons I can’t recall.

spardoko commented:
In regards to added configurable limits to vps allowing for overhead... Yes, the RAM and HDD are soft limits identified by text only.  The purpose of those limits are to let users know the upper and lower limits for each individual server regardless of how many resources are available within their plan.  

I also want to point out that if the web.config entries are missing then the behavior does not change.  In other words, this commit won't change the behavior for other SolidCP users unless they define the web.config entries.

When an account has one VM, this feature offers little difference other than overriding the max amount of CPUs available to assign to the VM and displaying the limits in text form.  The primary purpose for this feature is when there are multiple VMs per account.  The VMs can span multiple servers which is why I chose to put them in the web.config versus provider settings but I'd be happy to improve on it if a better way is suggested.

Unlike RAM and HDD, the CPU is a hard limit with this commit.  With the original behavior, assigning a number of CPUs per account applies to each VM within the account.  In other words, if I assign X CPUs to an account through addons then every VM within the account can have a total of X CPUs.  As an example, if I assign 8 CPUs to an account and the account has four VMs then each VM can have 8 CPUs for a total assignment of 32 CPUs.

With this commit, the amount of available CPU is decremented after it is assigned to a VM. As an example, if I assign 8 CPUs to account through addons and the account has four VMs then the total amount of CPU that can be assigned to all four VMs is 8 CPU.  If I assign 2 CPU to the first VM there will be a total of 6 available for assignment to the second VM.  If I assign 4 CPU to the second VM then there will only be 4 CPU available to split between the third and fourth VM. Of course, each VM will also be limited to a max number of CPUs defined in the web.config.  This will be true regardless of whether those VMs span multiple servers.

Unlike CPU, available RAM and HDD have always decremented with each new VM within an account.  That's partially why I left the per VM limits as soft limits for now.  I can make per VM limits hard limits sometime in the future.

Our VM plans within WHMCS are based on resources (CPU, RAM and HDD).  Our clients buy those resources individually.  If our client wants to buy two more procs to add to one of their servers then this commit allows them to do that.  Before this commit, there was no way to do that unless there was only one server in the account (CPU count applied to all servers).

We also don't want our clients to create VMs outside of the defined limits.  We allow them to buy lots of resources (CPU, RAM and HDD)because they can spread them out over lots of VMs.  However, we want to limit the amount of resources assigned per individual VM.

berkut1 commented:
So, is it possible to bypass the set quota? If that’s the case, maybe it’s better to fix that issue instead of creating a rather odd solution that doesn’t address the main problem—“quota bypass”?
Regarding creating space for servers for clients, ideally, it’s better to have one VM server = one hosting space. While it’s possible to create multiple servers within a single hosting space, this comes with certain limitations. For clients, it’s generally better to allocate a separate hosting space for each new server.

spardoko commented:
It's still limited to the quotas defined within the plan.  For a long time, we only deployed one server per hosting space but this has a lot of limitations too.

spardoko commented:
We also currently use admin only templates in production. For us, we mostly use it when importing VMs with OS's that we don't offer our clients for deployment.  And, for VMs that are currently deployed but we no longer want to make available for clients to deploy.

berkut1 commented:
For importing, it doesn't matter which OS you choose, it's just a value in the database. Also, if you stopped providing a certain OS, wouldn't it be easier to remove it? Or do you allow using an old/hidden OS for reinstallation? In that case, your solution doesn't solve the problem, and clients will be able to bypass the restrictions by reinstalling the server.

spardoko commented:
For importing, we still have to select an OS and while it may not matter from a functional perspective, the OS selection displays on the VM configuration page so it's nice to have it match the actual OS.

This feature is simply a way to differentiate between which images are available to admins versus clients.  It can be utilized in many different ways.

We deploy images as admins that we don't want to be available for self deployment by clients.  It could be older or other OS versions.  For example, we prefer to only allow clients to install the latest version of Windows server.  However, from time to time clients request an older version of Windows for whatever reason.  If we agree, we simply deploy it for them as the admin. 

We also retain some custom images for special services that we deploy and further customize as admins before we release to the client.  We mark these as admin only.

Admin only OS templates are a good solution for these and many other scenarios.  If I am missing something, please explain.

berkut1 commented:
Alright, but your implementation still doesn’t work for server reinstallation. Also, SolidCP has not only administrators but many other roles, such as CSR, Reseller, etc., yet you’ve strictly blocked access only for administrators. For example, in our case, the administrator role isn’t used at all (except for connecting and setting up new nodes); everything is handled through the CSR/Reseller role or its equivalent.

This also means the parameter name "Administrators only" isn’t very suitable here. Perhaps it would be better to use something like "Disabled/Hidden" instead.

And I really don’t like the solution for addressing the quota bypass issue.

spardoko commented:
It wasn't intended for non-administrators. It'll work for reinstallation if you're an administrator.  Hence, the name "Administrators Only."

I am aware of all the other roles.  I don't get your point.  It's intended for only Admins. If I would've allowed for other roles, I wouldn't have labeled it "Administrators Only."  It doesn't sound like you would want to use the feature even as a CSR/Reseller so I a don't plan to expand the feature to other roles. Or, maybe you realized it's a good feature and want to use it.  If so, just say that and I'll expand it to allow for more detailed selection of roles.

I am not going to label it "Disabled/Hidden" when it's not disabled or hidden for admins.

I imagine you don't like my solution for the quota bypass issue since I have not written any code for any such issue.  My code does not bypass any quotas.  It fixes a flawed implementation of CPU distribution and allows for per vm limits within the existing quotas - albeit soft limits.  If you don't like it, don't use it.

Quite frankly, I don't have a clue what you mean by quota bypass issue.  Based on your words, it seems like some kind of issue that you couldn't figure out how to solve.  If you want to share with me the details then I'll look into it and contribute what I can.

berkut1 commented:
I’m sorry, but the "if you don’t like it, don’t use it" approach is poor practice. You’ve provided a solution that doesn’t fix the problem but instead adds an optional extra layer that feels out of place and, most importantly, isn’t intuitive. Who will maintain its functionality in the future?

If you’ve found an issue with quotas, you should resolve it (or create an issue with a description and reproduction steps), rather than creating such an odd workaround that requires manual file editing. It looks like this was designed solely for your unique use case, ignoring all other users of this product.

Regarding the "Admin OS Template," I somewhat like the idea of hiding OS images from clients, but your implementation assumes that only administrators will handle tasks like reinstalling or creating servers. However, using an administrator for such minor tasks is unjustified and even risky due to the administrator’s extensive permissions. A simple human error (like clicking the wrong button) could lead to serious consequences.


Either way, I’ve shared my opinion; the final decision is up to the project maintainer.

FuseCP-TRobinson commented:
If there is a quota issue it should be fixed in the enterprise server not via a work around in the web portal. If its only fixed in the web then it can still be broken if using the api etc.

The idea of Hidden templates is a good idea but maybe this needs to be more Hosting plan based? Something like: Show hidden templates and then it can apply to resellers etc. If a serveradmin does load a customer it should always show these.

IvanSchulz commented:
Hi, we also use SolidCP with WHMCS and we haven't had any problems with CPU quotas so far. We packed all HyperVs into a virtual server, then created a hosting plan with it and added add-ons like additional vCores, RAM, HDD, etc. Can you describe how to reproduce the bug with CPU quotas?

spardoko commented:
Hi!  First of all, I agree with all of you that the configurable limits commit might not be the best approach.  I have been looking into how to make a better solution. I do believe the parameters should be removed from the web.config and integrated in a hosting plan policy with some parts being part of the provider and some with the enterprise server so that quotas can be enforced on API calls.  However, it has been working for me as it stands and solves many of my issues.  Let me explain the CPU quota issue in more detail...

In my SolidCP Hosting Plan for VMs, I set "Number of Servers" to unlimited.  Or, any number more than 1 will do. I set "Number of CPU cores" to 0.  

I have a SolidCP addon and set "Number of CPU cores" to 1.  I use configurable options in whmcs so that clients can purchase a desired number of CPU cores.

Let's say a client purchases 4 CPU cores through whmcs. Whmcs then provions the SolidCP addon with a quantity of 4. Now, the clients SolidCP Hosting space has unlimited servers and 4 CPU cores.  This is all fine so far.

The problem now is that the client can create as many VMs as they want with each VM having between 1 and 4 CPUs.  They could spin up 10 servers with each having 4 CPUs if they want.  If they go back into whmcs and increase the number of CPU cores from 4 to 8 then they can come back into SolidCP and edit the configuration of all 10 VMs and give each of the them between 1 and 8 CPU cores.

This commit changes that behavior.  It allows the client to provision only the exact number of CPU cores that they purchase.  So, for example, if the client purchases 4 CPU cores in whmcs and they spin up 1 server in SolidCP with 4 CPU cores then they will not be able to spin up any more servers until they buy more CPU cores in whmcs.  If the client goes back into whmcs and increases the number of CPU cores from 4 to 8 and then goes back to SolidCP to spin up another server then it will only allow them to add between 1 and 4 CPU cores to the new server because the first server is already using 4 of the 8 CPU cores that were purchased.

That's the first CPU issue. I hope my explanation makes sense.

Further, this commit addresses another problem related to CPU cores.  Currently, each VM is limited to the number of physical cores in one processor of the host server. At least that's how it seems to work for me. For example, if I have a physical host with 4 x 12 core procs with hyperthreading then the  number of CPU cores available to provision to any one VM in SolidCP on this host will be limited to 12.  Although the server itself will register 96 processors with hyperthreading. Hyper-V allows for oversubscription of CPU.  SolidCP does not.  I should be able to assign as many to CPU cores to a VM as I want rather than be limited to the number of physical cores by SolidCP, of course within the limits of the defined SolidCP quotas.  We monitor the utilization of resources in an entirely different system which is what I assume most of you do.

Let's say I create a VM in MSVMM with 48 processors on the host described above and then import the VM into SolidCP.  If you go into the configuration page of the VM after it has been imported into SolidCP it will display 48 CPU cores.  However, if you then click edit configuration, the dropdown for CPU cores will only display a maximum of 12 CPU cores which only truly becomes apparent after you click the dropdown.  If the client adds some additional RAM or something else to the VM and saves the configuration, the VM will immediately be knocked down from 48 to 12 CPU cores and it's not immediately apparent.  

This commit changes that behavior.  It allows the computed number of max CPU cores to be overridden by a value set by the administrator. In the example above, it allows me to provision the VM with 48 cores directly in SolidCP and, further, it will not automatically change the value when the configuration is edited.  Keep in mind that this example is only possible if the per vm limit is set to at least 48, the CPU core addon quota is at least a quantity of 48 and there are at least 48 unprovisioned CPUs in the hosting space.

In other words, the quotas in the plan still apply.  Setting the per VM limit to 48 as described in the example above will not change anything unless the clients CPU quota exceeds the number of physical CPU cores of one processor of their host.  If the number of CPU cores is set to 4 in the solidcp hosting plan then only 4 CPU cores will be available for provisioning as long as the per vm limit is set higher.

I hope this explanation makes sense too. This solves an issue that has been a real problem for me in the past.  We had to prevent clients with large amounts of CPU from editing their own configuration because SolidCP kept modifying the CPU core count on the VM whenever the client would save the configuration.  So when those clients wanted to increase RAM or HDD or whatever then we would have to do it manually in MSVMM and then execute a script to modify the SolidCP database with the new value.  Now, clients can modify their own VM configuration without concern for the number of CPU cores changing on their VM.

spardoko commented:
As for the admin only templates, I agree that the feature should be expanded to include other roles.  I would like to see the ability to create policies for OS templates and software images where roles, reinstall and other things can be defined for each and then applied to hosting plans.  However, that is more than I was willing to bite off when I added Admin only templates.  For me, I don't use the reseller role like berkut1 because the IP restrictions feature only applies to administrators and not resellers.  I'd like to see an optional IP restriction feature for other roles before expanding on templates and software. Please accept the commit as it stands now.  Although limited to administrators, still useful.

spardoko commented:
Ivan, I still need to review the resourcekey issue.  At least in English, the word update appears twice when the resource key exists.  Since I don't make German available, I didn't catch the problem with the dialogue.  So, I need to look into it more and figure out the correct solution.

berkut1 commented:
@spardoko 
Sorry, but I didn’t quite understand your explanation. I can’t check it right now because of the Chirstmas holidays, but I don’t remember SolidCP limiting the use of CPU cores in any way.
In my case, everything goes through a reseller who has plans created with Unlimited CPU initially specified. Then, Hosting Spaces are created for the reseller. In my setup, **Hosting Space = Node Hard Drive.** If there are two disk spaces on one node, **two Hosting Spaces** will be created for it. After that, each Hosting Space gets a separate plan with specific limits. For example, if I remember correctly, we set 32 CPU cores per plan (Our automatic system creates a server with 1-32 CPUs without users.).
Then, users are created under the reseller, and each of their Hosting Spaces will only have one server. When creating or editing a server, all CPU cores appear without any issues in the dropdown list.

Perhaps your issue only happens when there are more than two servers in one client Hosting Space? In our case, every user under the reseller has a separate Hosting Space for each server, and all of them can have servers with up to 32 CPU cores.

Maybe it would be better for you to show the issue using screenshots?

Regarding the "Admin OS Template," SolidCP is not your personal project. You should adapt changes to make them safe and useful for everyone, not just for yourself. In this case, as I mentioned earlier, the "Admin OS Template" is implemented in a way that is difficult to use and only covers your very specific use case.

spardoko commented:
@berkut1 It limits CPU based on GetProcessorCoresNumber() in the provider.  It shouldn't but it does. Changing that function or removing it entirely does not result in the same behavior as my solution so that was not an option.  Other than that, I don't how else to help you understand other than maybe consider its use beyond your own.

berkut1 commented:
Okay, now I understand where the problem is. You're talking about this:
https://github.com/FuseCP/SolidCP/blob/31119dbd958001b43d86e4a2783c7a6e779dc4dd/SolidCP/Sources/SolidCP.WebPortal/DesktopModules/SolidCP/VPS2012/VdcCreateServer.ascx.cs#L159

But if you’re saying that we can safely remove this check (and other if needed), then why did you do the opposite and create an additional layer? In your case, it would have been better to state that the check isn’t necessary and back it up with your tests.

berkut1 commented:
Also, regarding [allow underscore at front of domains](https://github.com/FuseCP/SolidCP/pull/218/commits/548344e43795361371c2a7c1657ec5063fb22ebb)
According to [RFC1035](https://datatracker.ietf.org/doc/html/rfc1035#section-2.3.1), underscores are not allowed.
However, they are permitted in DNS SRV records as specified in [RFC2782](https://datatracker.ietf.org/doc/html/rfc2782) for very specific use cases.

If your commit doesn’t distinguish between these two RFCs, then I’m against this commit as well. We shouldn’t break RFCs, as it could lead to unpredictable errors in the future. I recommend adding a condition for when the new rule can be applied.

FuseCP-TRobinson commented:
I cannot recreate the quota issue. When i make an VM it uses the limits and the same when i try edit it. Maybe this is a WHMCS plugin issue?

Hosting plan with addons for CPU etc
![image](https://github.com/user-attachments/assets/90fb39ff-15b6-47f9-b391-85bcc3d3894d)

Overview of VMs qith quotas:
![image](https://github.com/user-attachments/assets/daae8a50-8b72-43c2-86d1-02470063af89)

Editing a VM:
![image](https://github.com/user-attachments/assets/c8ab0dad-8e3c-4b7b-8c2d-4e5f46e7cd69)

When trying to add a new VM:
![image](https://github.com/user-attachments/assets/22c2100f-2829-4db3-a987-8ee69a874757)

@spardoko Let me know if i missed anything here on this issue and if you are seeing the same or not?

spardoko commented:
@berkut1 DNS/RFC is a complicated subject, no doubt.  Nonetheless, I believe your interpretation is wrong.  It's my understanding that the only restriction that applies to domains is the length.  The DNS is supposed to serve requests for any domain with any characters within the length restriction.  This is defined in [RFC 2181 Section 11 Name Syntax](https://datatracker.ietf.org/doc/html/rfc2181#section-11). This is different than the restrictions that you reference. DNS can be used for all sorts of things beyond simply serving publicly registered domain names that have restrictions on the types of characters that can be used.  I can understand why you would want to limit it in SolidCP to basically one use case as it relates to the host name. However, underscores are used in the wild all the time and there is nothing wrong with it.  

Further to that, I can serve up a zone named "_911._berkut1_" if I so desire.  According to RFC, the DNS must not refuse to serve this zone. It doesn't mean that others don't impose restrictions though.  If I added that domain to my DNS then you could and should be able to query it over the Internet albeit you would need to connect directly to my DNS since the root name servers will refuse it.

SolidCP already allows underscores when you add resource records.  Creating a zone labeled berkut1.com and adding an A record labeled _911 is currently allowed in SolidCP (at least for the latest MDDNS provider). So why would you care if we also allow a zone name to be created with the name _911.berkut1.com?  Other than the fact that you don't like me and want to be antagonistic.

I can appreciate the DNS name restrictions that SolidCP has in place because it covers most use cases.  If SolidCP were to allow a broad swath of characters then it should be limited to AD or private DNS only and differentiated as such.  However, simply allowing an underscore in a DNS domain violates no RFC spec. 

@FuseCP-TRobinson I'll try to recreate with screenshots.  Since we compile our own version, I'll need to push your v1.5.1 to an environment where I can reach the Hyper-V servers for testing.  This will take me some time.

FuseCP-TRobinson commented:
What would be the usecase for a _ in a domain name? While this is technically possible you wont get a SSL for it for example.

spardoko commented:
@FuseCP-TRobinson I've tried to recreate the same setup.  I'm not doing anything with whmcs in my setup. I get the same result as you. I would expect no more CPU cores available since you have reached the quota.  Can you change your addon to allow 48 procs and tell me what you see on the same screens?  Can you assign 44 processors to a third server after changing quota?  Can you limit the max cores on any one VM to something less than 44 at this point?  If so, please let me know how you're doing those things?

In my test, I never have more than two procs available.  On the host, I manually ran the same WMI command as the provider uses to get max cores and I return 64 logical processors. I can import a VM with up to 44 procs and it will work but if I edit and save configuration then it will get knocked down to 2. And, side note, even if I had more there is no way to limit the cores on one VM which is the title of the commit.

Hosting plan with addons for CPU etc (I changed to CPU to 48 after trying it with 4 procs like you)
![Screenshot 2024-12-28 142708](https://github.com/user-attachments/assets/7c4d0e47-fe75-4225-a59c-2908507dc9a7)

Overview of VMs qith quotas:
![Screenshot 2024-12-28 143029](https://github.com/user-attachments/assets/9d0d5eb2-03da-41de-a701-478ab3284c31)

Editing a VM:
![Screenshot 2024-12-28 143556](https://github.com/user-attachments/assets/a77d2fb1-3d9e-45b6-bd9e-d18091fe29aa)

When trying to add a new VM:
![Screenshot 2024-12-28 144026](https://github.com/user-attachments/assets/7503a920-15a9-43fa-b53f-141ac7231848)

Now I've applied my commit and continued on. When using my commit, nothing will change with regard to cpu unless you define the per vpn limits. After defining limits, it simply bypasses the maxcores calculation for CPU and use a maxcores static value instead.  It's a very simple change so it's not that big of a deal. This eliminates the calculated cap (which seems to work differently on different servers) and also does not permit any individual VM to have more than the static value of CPU (you can make the number very high if you desire). 

Settings for per VM limits in web.config (they apply globally)
![Screenshot 2024-12-28 150059](https://github.com/user-attachments/assets/eedc4135-40f3-402a-8dc3-d3628863122c)

Adding a third new VM
![Screenshot 2024-12-28 150605](https://github.com/user-attachments/assets/072df321-d092-4470-8b99-47fc2a400d92)

Overview of VMs with quotas after adding a third VM with max of 32 cores
![Screenshot 2024-12-28 152026](https://github.com/user-attachments/assets/9be2b7fc-0167-48d3-beaa-0fe53cfb521c)

Adding a fourth new VM (the plan quotas still apply)
![Screenshot 2024-12-28 152219](https://github.com/user-attachments/assets/c13a0878-e8e9-4a66-9d9e-a0d52a8fb2d0)

Overview of VMs with quotas after using up all resources
![Screenshot 2024-12-28 152833](https://github.com/user-attachments/assets/0c51d8c6-83ed-4427-b369-a20c7cbcb50d)

Creating a new VM, of course plan quotas still apply
![Screenshot 2024-12-28 152943](https://github.com/user-attachments/assets/e7d4f0d8-4019-4c2e-856b-08d9e926b12c)

spardoko commented:
@berkut1 Please see my screenshots. Can you please explain how I am using plans and spaces wrong?  If this is wrong, can you please provide details on how I should be doing it differently?

berkut1 commented:
@spardoko Please provide screenshots of the limitations in your plan from ServerAdmin to Reseller, which are linked to SCPUser. Specifically, how you set resource usage limits for the CPU.
There should be at least two plans one from Serveradmin, and second from Reseller.

spardoko commented:
@berkut1 Thanks.  Here are the screenshots. 

Serveradmin Plan
![Screenshot 2024-12-28 165750](https://github.com/user-attachments/assets/42692bf9-d24a-42cf-988d-10cf80060850)

Reseller Plan
![Screenshot 2024-12-28 165836](https://github.com/user-attachments/assets/b87faa0c-f4c3-499b-9088-78177a2e9185)

And, the CPU addon
![Screenshot 2024-12-28 165906](https://github.com/user-attachments/assets/31bbf2d9-b23e-4cc3-964c-937f4c0aef14)

berkut1 commented:
As I see it, the problem is that you set the CPU value for the Reseller to 0 and are then trying to increase it using addons.

If you set in the Reseller quota to 48 CPUs, you should be able to use them without any issues. However, it seems you want to extend this through addons. In that case, I suspect you should start with at least 1, not 0.

There might also be an issue similar to the one I encountered here: https://github.com/FuseCP/SolidCP/pull/218#issuecomment-2561221868. It’s possible that when you add an addon, it doesn’t automatically increase the quota. Go to the client’s Hosting Space settings and try manually editing the values there to increase the limits after adding the addon.

spardoko commented:
@FuseCP-TRobinson What would be the usecase for a _ in a domain name? While this is technically possible you wont get a SSL for it for example.

Answer: Delegation of DNS

Example:  My client may want to allow certain employees of their business to manage specific subdomains.  I can accomplish this, for the most part, by creating a reseller account with DNS service.  I add the domain and permit subdomains.  Then I create user accounts for each person that manages those subdomains and add the subdomain as a domain within that account.  Those subdomains might also include domainkey/dkim records with underscores.  

Keep in mind that I can create the subdomain with an underscore directly in DNS.  Because it's a legitimate thing to do.  I can also import into SolidCP as a domain with an underscore.  That's how I would add these records if I didn;t have my commit to allow the underscore.

@berkut1 Please do not respond to this.  I DO NOT CARE what you think about how I use this feature.

spardoko commented:
@berkut1 I tried removing the CPU addon and overrode the plan settings to specify 48 CPUs and still only get 2 in the dropdown.

With the addon deleted, I turned the quotas back to the plan settings and went into then specified 48 CPUs directly in the hosting plan.  I still get 2 CPU in the dropdown.

Here is the plan with 48 CPUs
![Screenshot 2024-12-28 173403](https://github.com/user-attachments/assets/3fa70f85-fba3-4bef-a4c3-98408316e5eb)

Here are the user quotas.  I have removed CPU addon.
![Screenshot 2024-12-28 173457](https://github.com/user-attachments/assets/a4216528-85c7-4073-8087-126d7fc1b154)

I deleted the fourth server so we have 12 CPU free now.
![Screenshot 2024-12-28 173529](https://github.com/user-attachments/assets/183a5810-6d0d-4b82-a600-b0fd773f83ac)

I create a new VM and get 2 CPU
![Screenshot 2024-12-28 173609](https://github.com/user-attachments/assets/004e4806-43e5-4231-b43d-3bd805df4345)

I can get a different number in the dropdown if I change hosts that uses different kinds of processors.

spardoko commented:
@FuseCP-TRobinson Can you please try to recreate my steps in your test account?  I am curious to know what you experience.  Please change the CPU from 4 to 48 and tell me if you can create the additional VMs as I did.  Thanks

spardoko commented:
Sorry, I did not mean to close the pull request!

berkut1 commented:
Alright, this seems strange if you actually removed the addon and changed the CPU value for the reseller from 0 to 48.

Maybe the value got cached in the database and wasn’t updated? Try deleting the existing Hosting Space and creating a new one with the updated conditions. Alternatively, remove the plan with the Hosting Space (for Reseller) and recreate them, setting the CPU value to 48 right from the start.

spardoko commented:
@berkut1 Okay, I'll try it and let you know but I am done for today.  I've got to say though, I don't believe that the quota doesn't automatically increase when you add an addon.  The quota seems to be returning the right number with the addon.  From what I can tell, it's only the maxcores calculated value that's the problem. It is limiting the quotas. Removing it allows the quotas to take full effect.  Further, you still need to specify a maxcores  value otherwise... how would you populate the CPU dropdown if unlimited CPUs is selected in the hosting plan or prevent a massive number of CPUs from being added to one VM.  My solution seems to be a really good fix for this issue.  But again, I'm all ears if you can come up with a better solution.

berkut1 commented:
Unfortunately, your solution is not suitable for the reasons mentioned earlier. As it has already been determined (at least in our case, there are no issues without addons), the issue lies with the plans and addons, and this needs to be fixed. There are no other options here. If you cannot fix it, then you should create an issue and hope that one of us will have the time to address the problem.

berkut1 commented:
I checked the code, and I think the problem lies in this condition:
https://github.com/FuseCP/SolidCP/blob/31119dbd958001b43d86e4a2783c7a6e779dc4dd/SolidCP/Sources/SolidCP.WebPortal/DesktopModules/SolidCP/VPS2012/VdcCreateServer.ascx.cs#L172-L178

Let’s suppose that your maxCores is 2 CPUs:
https://github.com/FuseCP/SolidCP/blob/31119dbd958001b43d86e4a2783c7a6e779dc4dd/SolidCP/Sources/SolidCP.WebPortal/DesktopModules/SolidCP/VPS2012/VdcCreateServer.ascx.cs#L159

As I see, your CPU quota is 36 out of 48.
In this check:
https://github.com/FuseCP/SolidCP/blob/31119dbd958001b43d86e4a2783c7a6e779dc4dd/SolidCP/Sources/SolidCP.WebPortal/DesktopModules/SolidCP/VPS2012/VdcCreateServer.ascx.cs#L170
we will get true (48 > 36).

And in this check:
https://github.com/FuseCP/SolidCP/blob/31119dbd958001b43d86e4a2783c7a6e779dc4dd/SolidCP/Sources/SolidCP.WebPortal/DesktopModules/SolidCP/VPS2012/VdcCreateServer.ascx.cs#L172

48 + 1 - 36 = 13 > 2
This means that in your situation, we will always fall into this check:
https://github.com/FuseCP/SolidCP/blob/31119dbd958001b43d86e4a2783c7a6e779dc4dd/SolidCP/Sources/SolidCP.WebPortal/DesktopModules/SolidCP/VPS2012/VdcCreateServer.ascx.cs#L172-L178

As a result, you will always get 2 CPUs.

If my thoughts are correct, then this situation needs to be fixed. We cannot remove maxCores because it would break backward compatibility in cases where the maximum value is not specified in the plan. This means we need to add a check here:
https://github.com/FuseCP/SolidCP/blob/31119dbd958001b43d86e4a2783c7a6e779dc4dd/SolidCP/Sources/SolidCP.WebPortal/DesktopModules/SolidCP/VPS2012/VdcCreateServer.ascx.cs#L172

We need to ensure that a CPU value is specified in the quota, and if it is not set to unlimited, then execute this code:
https://github.com/FuseCP/SolidCP/blob/31119dbd958001b43d86e4a2783c7a6e779dc4dd/SolidCP/Sources/SolidCP.WebPortal/DesktopModules/SolidCP/VPS2012/VdcCreateServer.ascx.cs#L179-L185

I assume we need to first check if the quota limit value is greater than the physical cores, for example `isQuotaAllocated`:
```c#
            // bind CPU cores
            int maxCores = ES.Services.VPS2012.GetMaximumCpuCoresNumber(PanelSecurity.PackageId);

            QuotaValueInfo cpuQuota2 = cntx.Quotas[Quotas.VPS2012_CPU_NUMBER];
			
	     bool isQuotaAllocated = cpuQuota2.QuotaAllocatedValue > 0 

```

Then we can simply do this `if (!isQuotaAllocated && ((cpuQuota2.QuotaAllocatedValue + 1 - cpuQuota2.QuotaUsedValue) > maxCores))`:
```c#
                        if (cpuQuota2.QuotaAllocatedValue == -1)
			{
				for (int i = 1; i < maxCores + 1; i++)
					ddlCpu.Items.Add(i.ToString());

				ddlCpu.SelectedIndex = ddlCpu.Items.Count - 1; // select last (maximum) item
			}
			else if (cpuQuota2.QuotaAllocatedValue >= cpuQuota2.QuotaUsedValue)
			{
				if (!isQuotaAllocated && ((cpuQuota2.QuotaAllocatedValue + 1 - cpuQuota2.QuotaUsedValue) > maxCores))
				{
					for (int i = 1; i < maxCores + 1; i++)
						ddlCpu.Items.Add(i.ToString());

					ddlCpu.SelectedIndex = ddlCpu.Items.Count - 1; // select last (maximum) item
				}
				else
				{
					for (int i = 1; i < (cpuQuota2.QuotaAllocatedValue - cpuQuota2.QuotaUsedValue) + 1; i++)
						ddlCpu.Items.Add(i.ToString());

					ddlCpu.SelectedIndex = ddlCpu.Items.Count - 1; // select last (maximum) item
				}
			}
			else
			{
				ddlCpu.Items.Add("0");

			}
```

I'm not 100% sure about my solution; the fix was written "on the fly."

spardoko commented:
Your suggested change may fix some of the problem but mine offers a solution for a much bigger picture.  I actually don't think that the maxcores computed value should be overridden simply because there's unallocated quota in the plan.

My code already modifies the block of code that you are referencing in both vdccreateserver.ascx.cs and vpsdetailseditconfiguration.ascx.cs.  It also continues to use the maxcores calculation for backwards compatibility.  It simply adds the ability to override maxcores with a static value from the web.config if you so desire.  I rewrote the block to simplify the conditional statements so that they are easier to follow, otherwise the result didn't change other than the ability to optionally use a static value.

For me, the only question is whether a static maxcores value should reside in the web.config or someplace else and whether other users would want the labeling on the configuration page.

I chose the web.config because I needed it to be absolutely consistent across the board including across plans and I couldn;t think of any better place to put it at the time.  I have over 500 physical servers dedicated to Virtualization.  VMs are migrated to different hardware all the time.  I can't take the chance that maxcores can change during these backend operations which can ultimately change the VM CPU allocation when a user edits the configuration.

Using a static value keeps things consistent and avoids a whole bunch of issues especially when you have a lot of host servers.   It allows you to use the hosting plans however you want and remain consistent - using addons, allowing unlimited CPU, etc.  It also allows you to cap the number of CPU cores assigned to individual VMs (in a multi VM plan) to a value less than the hosting plan quota - which is something not possible before or with your suggested solution.  Hence, the title of the commit.

berkut1 commented:
What you're trying to do is forcibly bend SolidCP to fit your specific use cases. In your case, you should be writing your own separate layer or middleware outside of SolidCP that operates through the API to enforce your conditions, fixing the bugs that arise in the process, such as the one with maxCores.

If you're dealing with complex migrations that SolidCP can't handle, then you should either improve SolidCP itself or, at the very least, propose ideas for its enhancement.

One way or another, it seems that you are determined to stick to your position, so I have nothing more to add here.

FuseCP-TRobinson commented:
When i use a reseller that seems to cause the problems. I can see the quota shows:

![image](https://github.com/user-attachments/assets/2396cccc-4701-4aa8-b20b-0374d2b9c2b6)

But when i try add a new VM:

![image](https://github.com/user-attachments/assets/6a752663-5b14-478f-872d-c1a0d3d779b0)

For this we need to check into a quota.

However from looking at your commit your looking at a way to set a max cores per VM limit? This should be done in Policies and not via a web.config setting (All Web.config settings will be removed in the upcoming CoreWCF version).

berkut1 commented:
@FuseCP-TRobinson 
Can you check if my fix works? If it does, I think it makes sense to use it since @spardoko doesn't plan to fix the bug.

FuseCP-TRobinson commented:
> @FuseCP-TRobinson Can you check if my fix works? If it does, I think it makes sense to use it since @spardoko doesn't plan to fix the bug.

Yep this worked on creating a new VM i can select all 48 cores.

berkut1 commented:
@FuseCP-TRobinson Thanks, also, the same needs to be done in [VpsDetailsEditConfiguration.ascx.cs](https://github.com/FuseCP/SolidCP/blob/master/SolidCP/Sources/SolidCP.WebPortal/DesktopModules/SolidCP/VPS2012/VpsDetailsEditConfiguration.ascx.cs)

spardoko commented:
@FuseCP-TRobinson I agree that the setting should be made in a policy rather than web.config.

I don't agree with the fix that you guys made though.  It does not solve the problem.  It only solves the one issue you tested for.  It literally just bypasses the maxcores calculation when there is unused quota.

FuseCP-TRobinson commented:
I agree it fixes the issue but causes another one where maxCores is now not taken into account which means if a user is to set it above the allowed cores the VM wont start with no clear reason as to why.

berkut1 commented:
Let’s try to summarize what you are trying to achieve.
You’re saying that you have a server with 4 processors, each with 12 threads, which should total 48 threads, but it only shows 12?

But doesn’t this code count them correctly?
https://github.com/FuseCP/SolidCP/blob/3de665b7bbb7ccd75214ec38d082f1ec6efc9a80/SolidCP/Sources/SolidCP.Providers.Virtualization.HyperV-2012R2/HyperV2012R2.cs#L2146-L2154

From the code, I can see that it loops through each processor and sums up all their threads. Isn’t that happening in your case?

If not, then maybe my fix really isn’t suitable (and perhaps the changes should be reverted), but we still need to figure out why it’s not summing up all CPU threads in your case.

FuseCP-TRobinson commented:
I think for VpsDetailsEditConfiguration.ascx.cs we should set this to:

`                // bind CPU cores
                int maxCores = ES.Services.VPS2012.GetMaximumCpuCoresNumber(vm.PackageId);
                PackageContext cntx = PackagesHelper.GetCachedPackageContext(PanelSecurity.PackageId);

                QuotaValueInfo cpuQuota2 = cntx.Quotas[Quotas.VPS2012_CPU_NUMBER];
                int cpuQuotausable = (cpuQuota2.QuotaAllocatedValue - cpuQuota2.QuotaUsedValue) + vm.CpuCores;

                if (cpuQuotausable > maxCores)
                {
                    cpuQuotausable = maxCores;
                }

                if (cpuQuota2.QuotaAllocatedValue == -1)
                {
                    for (int i = 1; i < maxCores + 1; i++)
                        ddlCpu.Items.Add(i.ToString());

                    ddlCpu.SelectedIndex = ddlCpu.Items.Count - 1; // select last (maximum) item
                }
                else if (cpuQuota2.QuotaAllocatedValue >= cpuQuota2.QuotaUsedValue)
                {
                    if (cpuQuotausable > maxCores)
                    {
                        for (int i = 1; i < maxCores + 1; i++)
                            ddlCpu.Items.Add(i.ToString());

                        ddlCpu.SelectedIndex = ddlCpu.Items.Count - 1; // select last (maximum) item
                    }
                    else
                    {
                        for (int i = 1; i < cpuQuotausable + 1; i++)
                            ddlCpu.Items.Add(i.ToString());

                        ddlCpu.SelectedIndex = ddlCpu.Items.Count - 1; // select last (maximum) item
                    }
                }
                else
                {
                    for (int i = 1; i < vm.CpuCores + 1; i++)
                        ddlCpu.Items.Add(i.ToString());

                    ddlCpu.SelectedIndex = ddlCpu.Items.Count - 1; // select last (maximum) item

                }`

This will make sure the drop down is limited to maxCores

berkut1 commented:
@FuseCP-TRobinson  
The problem is that we’re playing a game of broken telephone and wandering in the dark. We need detailed information about where exactly @spardoko is facing issues and how he’s using SolidCP. So far, we’ve figured out that `maxCores` plays an important role, and we can’t ignore it.

FuseCP-TRobinson commented:
@berkut1 My last comment was based on my test machine which is limited to 10 CPUs and max cores is ignored. With the changes i made it fixed that issue to where it limis the drop down to 10CPUs when an quota is above the maxCores.

I can also see on another server its 40 cores total while the drop down still shows 48. 

```
PS C:\Users\Administrator> Get-WmiObject -Class Win32_Processor | Format-List -Property NumberOfLogicalProcessors

NumberOfLogicalProcessors : 20
NumberOfLogicalProcessors : 20
```

berkut1 commented:
@FuseCP-TRobinson 
Yeah, I suppose we should revert it to:
```c#
// bind CPU cores
int maxCores = ES.Services.VPS2012.GetMaximumCpuCoresNumber(vm.PackageId);
PackageContext cntx = PackagesHelper.GetCachedPackageContext(PanelSecurity.PackageId);

QuotaValueInfo cpuQuota2 = cntx.Quotas[Quotas.VPS2012_CPU_NUMBER];
int cpuQuotausable = (cpuQuota2.QuotaAllocatedValue - cpuQuota2.QuotaUsedValue) + vm.CpuCores;

if (cpuQuota2.QuotaAllocatedValue == -1)
{
	for (int i = 1; i < maxCores + 1; i++)
		ddlCpu.Items.Add(i.ToString());

	ddlCpu.SelectedIndex = ddlCpu.Items.Count - 1; // select last (maximum) item
}
else if (cpuQuota2.QuotaAllocatedValue >= cpuQuota2.QuotaUsedValue)
{
	if (cpuQuotausable > maxCores)
	{
		for (int i = 1; i < maxCores + 1; i++)
			ddlCpu.Items.Add(i.ToString());

		ddlCpu.SelectedIndex = ddlCpu.Items.Count - 1; // select last (maximum) item
	}
	else
	{
		for (int i = 1; i < cpuQuotausable + 1; i++)
			ddlCpu.Items.Add(i.ToString());

		ddlCpu.SelectedIndex = ddlCpu.Items.Count - 1; // select last (maximum) item
	}
}
else
{
	for (int i = 1; i < vm.CpuCores + 1; i++)
		ddlCpu.Items.Add(i.ToString());

	ddlCpu.SelectedIndex = ddlCpu.Items.Count - 1; // select last (maximum) item

}
```
If `cpuQuotaUsable` > `maxCores`, then we provide the dropdown list based only on `maxCores`.

spardoko commented:
@berkut1 @FuseCP-TRobinson On the host I used in the example, it has two physical processors.  I had run the WMI command on it before and it produced 64 logical processors.  Give me some time and I can try it on bunch of different servers with different procs including one with 4 physical procs. I have not tried the second powershell command but I will try that as well.

I can tell you that I've seen anywhere between 2 and 12 cores on different physical hosts, at the most, in the CPU dropdown.  Never more than 12.  I can tell you that every single one of my servers has at least 48 logical processors.  Therefore, when using the commit, I set the static value to 48 in my SolidCP and it solves all the issues.  People would need to sort of figure out that number based on their environment if they want to solve all issues with CPU allocation.

@berkut1 I completely understand where at with this regarding fixing the bug and I don't disagree with you.  The problem is that the only way I could solve the bug was to define a per vm limit.

Keep in mind, even if you get maxcores value to match the right logical processor count it is still a problem.  This is because you might migrate the VM to a host that has a different logical processor count. What do you do then?

FuseCP-TRobinson commented:
I reverted the changes. Now i see the following:

![image](https://github.com/user-attachments/assets/d4ed1373-b361-42a5-9285-40f4acbd3add)

![image](https://github.com/user-attachments/assets/9ae0c63b-49c5-4a9d-b872-73d14d2c131a)

I think i have found the problem. The maxCores is fetched from the machine running SolidCP Server not the remote HyperV server.

berkut1 commented:
@FuseCP-TRobinson Yeah, if someone uses Remove Hyper-V instead of installing the SolidCP server on each server, they will get the wrong result.

berkut1 commented:
@spardoko 

> Keep in mind, even if you get maxcores value to match the right logical processor count it is still a problem. This is because you might migrate the VM to a host that has a different logical processor count. What do you do then?


There won't be any issues if the new host has the same or more CPU cores after we fix the issue with maxCores.
However:
If you're doing this with SolidCP, then you should adapt the code so that quotas are updated, or prevent migrations altogether if there is a quota overage.
If you're doing this without SolidCP (using external scripts), then I see the problem in your method, not in SolidCP. In this case, you need to write a separate system that will synchronize data in SolidCP, since you're bypassing SolidCP.

IvanSchulz commented:
Hi @berkut1 , we previously had the problem that if a VM was moved to another HV (by Cluster Manager in our case), SolidCP could no longer find and control it. I have made the function in SolidCP, which then reassigns the VM back to the correct HyperV server, but nothing is done with quotas. As I said, we also use SolidCP with WHMCS (a hosting plan is associated with WHMCS product and SolidCP addons with WHMCS configurable options) and we have no problems with quotas so far - it gets maximum vCores quota (hosting plan + addons), either maximum available HV cores displayed. We installed the SolidCP server on all HyperV nodes, then packed all SolidCP servers into a virtual server and we use the virtual server for hosting plans. We didn't have a scenario where a VM could have more vCores than an HV node, since we move a VM either within the cluster (where all HVs have the same processors) or from the old to the new cluster (where all HVs have more cores).

berkut1 commented:
@spardoko If you want, you can try testing this fix for `maxCores`.
https://github.com/FuseCP/SolidCP/issues/219#issuecomment-2565261149


I currently can't test this on a remote server and have only tested it locally with one processor. In theory, it should work.

spardoko commented:
@berkut1 @FuseCP-TRobinson I did not mean to close this pull request again.  Sorry about that. Let me try to fix.

spardoko commented:
@berkut1 This makes a lot of sense.  We use remote server settings.  We point the SolidCP Server URL to a high availability cluster of VMs whose purpose is to proxy the hyper-v connections.  Those all have the SolidCP server component installed.  None of our Hyper-V servers have the SolidCP server component.

Currently, all of the VMs in the cluster have 2 procs which is why I only ever see two procs right now.  It hasn't always been setup this way though.  The varying number of CPU I was seeing in the past was either when it was pointing to another url/server or it landed on a VM in the cluster with a different number of CPUs.

I believe your fix will work but testing will take some work on my end.  I'll test and let you know if I find any other problems.

We still want a feature to define per vm limits, including CPU, but I will look at putting that into a policy and making it better.

berkut1 commented:
@spardoko Create (different?) limits for each server within a single Hosting Space is, I think, very difficult to implement adequately because limits are tied exclusively to the Hosting Space, not to objects within it. You might explore ideas where a client within one Hosting Space could have child Hosting Spaces, similar to how Resellers work. Following this approach, with one server per Hosting Space, you could manage limits for each server within a single parent Hosting Space for the client.

This implementation - making the client function like a Reseller—seems the simplest and most logical solution. Much of the functionality already exists for Resellers, so it would just require extending it to clients and ensuring that it doesn’t break anything else.

P.S After writing these thoughts, it seems to me that this is the most correct solution, and you should consider exploring this direction.

spardoko commented:
@berkut1 No, not at all, I don't mean different limits for each server within a single hosting space.  I mean have the ability to create a policy for RAM, HDD and CPU that would apply to all servers within a space. 

First, keep in mind that we want to have hosting plans as I've described before.  Multiple VMs in one plan and use addons for CPU, RAM and HDD that can be applied by our clients however they want.  Similar to the screenshots.  Clients would be able to buy a bunch of CPU, RAM and HDD and then launch new and edit server configurations however they want and if we apply a policy then they will be forced to follow those rules.

Within that space, we may want to cap the maximum CPUs per VM to some number less than maxcores. The quota may say the plan has 100 CPU but the policy may say only apply 10 CPU can be applied to a single VM.  And maybe we only want to allow clients to provision RAM in 512 MB increments. And maybe we want to cap HDD on any one VM to 1TB even though the plan quota might be 10 TB.  And maybe we want to only allow HDD to be provisioned in 25 GB increments. Or whatever.  In any case, plan quotas would still apply. We also want some way to show clients what resources are available and what policies are in place when provisioning/editing - this is especially true for RAM and HDD since they are not a dropdown.  Although, if there was a policy in place, I suppose they could be dropdowns too (just thinking out loud).

We can accomplish some of that by putting one VM in one space but not all.  Unless I am missing something with the one server approach, and please apprise me if I am, clients won't have the ability to share private vlans between servers, be able to switch external IPs from one server to another, are not be able to reallocate resources like CPU and RAM to other servers, etc etc.  Those things are possible in a multi VM plan.  In any case, it's certainly more user friendly to have all the VMs for a client/space/usegroup whatever all on one page.

We actually had a one server per space approach in the past.  It simply has not worked for us (and our clients).

I appreciate your feedback.

berkut1 commented:
Alright, I think I vaguely understand your idea. You’re allocating resources to clients but also want to control how they use those resources within their quota.

If you can integrate this into SolidCP in a clean way that doesn’t burden the codebase (for example, the web.config approach is a bad example of how not to do it), then go ahead and give it a try.

However, in your case, I’d approach it differently. I’d create a separate website and manage all SolidCP resources exclusively through its API. This way, clients would have their own portal with specific rules, like the resource management you described, while SolidCP’s API would handle quota checks.

If you’re interested, I’ve built a wrapper for the SolidCP API for my company to simplify automation. https://github.com/berkut1/scpm
The idea is simple: we have predefined server plans (a common practice for server rental companies), and we just send those plans to the API, which handles everything - choosing where to deploy and how much resources to allocate. Clients only get access to basic server management in the panel, like rebooting, reinstalling, or changing passwords.

spardoko commented:
"You’re allocating resources to clients but also want to control how they use those resources within their quota."

Yes, exactly.

"If you can integrate this into SolidCP in a clean way that doesn’t burden the codebase (for example, the web.config approach is a bad example of how not to do it), then go ahead and give it a try."

You don't need to keep beating me up over the approach.  I've already agreed the web.config is bad and that I would need to do better.

"However, in your case, I’d approach it differently. I’d create a separate website and manage all SolidCP resources exclusively through its API. "

Maybe I don't understand but this doesn't seem like a good approach to me.  I mean I see the value in what you've created. I just don't think it's right for me.  It seems like a lot of work to simply be able to have some control over how clients provision vm resources.  Also, seems like I would need to reinvent much of what SolidCP already does.  I'd prefer to throw a bunch of resources into SolidCP and give clients the flexibility to use it however they want.  I've been using this project in some form or another since the dotnetpanel days.  I thought that's how it was intended to be used.

berkut1 commented:
If you have the determination to do it, then give it a try.

> Maybe I don't understand but this doesn't seem like a good approach to me. I mean I see the value in what you've created. I just don't think it's right for me. It seems like a lot of work to simply be able to have some control over how clients provision vm resources. Also, seems like I would need to reinvent much of what SolidCP already does. 

As for your concern that creating a separate management layer seems complicated - after you spend time trying to properly integrate your ideas into SolidCP without breaking anything else, you’ll know SolidCP so well that building a standalone application using its API will feel like a piece of cake.:)

spardoko commented:
If we communicating with each other you will probably eventually learn how to be a good ambassador to this project rather than just a good programmer.  Maybe then people will be inspired to want to contribute to this project. :)

spardoko commented:
@berku1 My apologies for my comments. I genuinely appreciate your contributions to this project.

FuseCPDevOPS commented:
Closing due to inactivity.
